### ISSUE TITLE
Game takes forever to start with a high number of songs

### ISSUE THREAD

**Issue by barbeque-squared** [UltraStar-Deluxe/USDX#685]
When trying to reproduce #680 I generated 30k test 'songs', all by the same 'artist' but with different titles. The problem is the following:

Loading screen time with ~3000 songs: ~5 seconds
Loading screen time with ~30k songs: I killed it after >3 minutes

None of the 30k test songs generate an error in Error.log (thus ruling out some IO/buffer problem in there)

Expected behaviour: the loading screen time is about linear in the number of songs
Actual behaviour: this doesn't feel linear. It doesn't feel like _O(n log n)_ either.

Tested with most recent git version on Archlinux

**Comment by barbeque-squared** [comment]
I'm going to suspect this has something to do with songs being created/analyzed from filenames instead of file object pointer things. I think this is fine for later (ie if a harddisk disconnects midgame, it should probably fail in a nice way) but if that happens _on the loading screen_ I don't think we need to be super nice about it.

I'll make an attempt to make the USong constructor take the object instead of a path (because the path can then be gotten from the object anyway) and see what that does.

**Comment by barbeque-squared** [comment]
Answering my own question: these 4 lines in `USong.pas` `ReadTXTHeader` are the culprit:
```
if self.Cover.IsUnset then
  self.Cover := FindSongFile(Path, '*[CO].jpg');

if self.Background.IsUnset then
  self.Background := FindSongFile(Path, '*[BG].jpg');
```

Okay, the situation here is fairly unique because there's about 30000 files in this directory, which isn't something that normally happens. Although I do have to ask: what's the point if this "feature" anyway? As in: you can set cover/background in the txt (or not), but why does it need to try and do something "smart"?

The one for the cover has always been there, but the one for the background was added by @bohning in 7f24cb9. But the commit message doesn't tell me anything. Is this (still) something that we should be doing? Or can we remove this bit of undocumented magic for txt's?

Because although it barely makes any difference on my actual song catalog, I don't like this kind of undocumented magic either.

**Comment by s09bQ5** [comment]
Shouldn't it be easy to cache the result of FindSongFile? It is called with three different file name patterns.

Or do we parse txt files in random order?

**Comment by barbeque-squared** [comment]
Chatted with bohning, apparently it is (or was) a fairly standard naming convention back in the day, and to be fair, I don't think regular users put 30000 txt's (all of which miss both the background and the cover) in one single folder anyway.

Having them all in one directory was just easy to generate for testing purposes, but it's not realistic. The problem [should] for the time being also go away if I just assign them all some generated cover(s) and background.

Caching would solve this problem entirely. txt's that are in the same folder are all handled consecutively. the _order_ might still be random, but unless there's also a bunch of subfolders, they'll always be together so you'd only need to cache a single path/folder at a time (look at `BrowseTXTFiles` and `FindFilesByExtension` in `USongs`).

**Comment by dgruss** [comment]
on windows 11 and around 28k txts but no covers or background images:
disabled via config.ini: real    0m25.402s
search result cached: real    0m26.219s
current master: real    0m27.374s

is this much worse when covers or background images are present in the directories?

i would also be in favor of removing the feature altogether. if i don't want a song to have a cover / background then the game shouldn't just auto-assign something else from that folder.

**Comment by dgruss** [comment]
+ current master: real 0m27.374s
+ search result cached: real 0m26.219s
+ disabled via config.ini: real 0m25.402s
+ with MD5 not computed on loading: real    0m22.235s
+ skipping the parsing of the note part of the songs would get it down to ~14s from here but then the note part of the song would not be checked on launch

**Comment by barbeque-squared** [comment]
> is this much worse when covers or background images are present in the directories?

This is what I used [00-generate.sh](https://github.com/user-attachments/files/24169921/00-generate.sh). Comment out the COVER and BACKGROUND lines and watch the loading time explode. And then be totally fine again if you comment out the code lines in https://github.com/UltraStar-Deluxe/USDX/issues/685#issuecomment-1546729846.

Basically what's happening is that that `FindSongFile` call is looping through 30.000 (plus a handful) of files _for every txt_ but it's a completely unrealistic folder structure. In a normal setup you'd have 30.000 _folders_ with each only a handful of files.

30.000 * 30.000 is (a lot) bigger than 30.000 * 5.

I don't know if caching handles this -- I'd also be fine just yeeting this "functionality" altogether unless other karaoke programs are also doing this, and even then I'd argue that https://github.com/bohning/usdb_syncer/issues/455 would have been a saner naming scheme regardless.

**Comment by dgruss** [comment]
i have fewer folders as i have 1 per artist, so a bit slower on my end maybe...
I'm fine with yeeting the cover + background matching functionality as well, in particular because there are people who don't want this to do this search anyways. i intentionally only have covers on very few txts (as a quality indicator) and background for none.

**Comment by basisbit** [comment]
I vote strongly against removing support for `[CO].jpg` and `[BG].jpg`. I know lots of setups that rely on it.
Instead, feel free to improve performance by for example caching the result for the check for a specific folder.

**Comment by barbeque-squared** [comment]
I've done a bit of searching and I can find _only two_ commits where this "feature" appears to be "introduced":

**Commit 1: "oops sorry I forgot this one .."** (Jan 13, 2008)
https://github.com/UltraStar-Deluxe/legacy-sourceforge-svn-mirror/commit/8226572159609fc0ee15de155d84576c00163371 (search for `self.Cover`)

And then cheekily `bugger !` on the second line of the commit message. The surrounding commits don't tell me _anything_ remotely useful either.

**Commit 2: USDX is now able to load SingStar XML Songfiles** (Mar 27, 2008)
https://github.com/UltraStar-Deluxe/legacy-sourceforge-svn-mirror/commit/f03ed07bc00ddc28a77f4e81d25e7d3ee596e06d (search for `self.Cover`)

See #816. We don't even support those XML files anymore, and the automatic idk what the singstar xml conversion command was iirc always added the appropriate headers (at least the cover one, did singstar even come with static backgrounds anyway?)

**In short**, the background loading appears to be _solely_ there for singstar XML files, which we stopped supporting almost two years ago. The cover one was slightly earlier, but there's absolutely zero context for _why_ it needed to be added.

Unlike basisbit I know absolutely no one that relies on this "feature".  Performous also appears to have it in a way more extended form, but its first iteration https://github.com/performous/performous/commit/e3f54d22f711e9b01a4025d377d517400203229a is from long _after_ USDX had the `[CO].jpg` behaviour.

**Comment by dgruss** [comment]
can be closed
